File: admin/cockpit/fleet-commander-admin/js/fcdbusclient.js
/*
 * Copyright (C) 2014 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the licence, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 * Authors: Alberto Ruiz <aruiz@redhat.com>
 *          Oliver Gutiérrez <ogutierrez@redhat.com>
 */

function FleetCommanderDbusClient(readycb, errorcb) {
    var self = this;
    var errorhandler = errorcb || function (err) {
        console.log('FC: Error - ' + err);
    };

    this._service = cockpit.dbus('org.freedesktop.FleetCommander', { bus: 'session' });
    this._proxy = this._service.proxy();

    this._proxy.wait().done(function (resp) {
        console.log('FC: Dbus service loaded');
        readycb(resp);
    }).fail(function (err) {
        console.log('FC: Failed to connect to Dbus service');
        errorhandler(err);
    });

    // Initialization methods
    this.GetInitialValues = function (cb) {
        self._proxy.GetInitialValues().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.DoDomainConnection = function (cb) {
        self._proxy.DoDomainConnection().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.HeartBeat = function (cb) {
        self._proxy.HeartBeat().done(
            function (resp) {
                cb(resp);
            }
        ).fail(errorhandler);
    };

    // Hypervisor configuration methods
    this.GetHypervisorConfig = function (cb) {
        self._proxy.GetHypervisorConfig().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.CheckHypervisorConfig = function (data, cb) {
        self._proxy.CheckHypervisorConfig(JSON.stringify(data)).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.SetHypervisorConfig = function (data, cb) {
        self._proxy.SetHypervisorConfig(JSON.stringify(data)).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.CheckKnownHost = function (hostname, cb) {
        self._proxy.CheckKnownHost(hostname).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.AddKnownHost = function (hostname, cb) {
        self._proxy.AddKnownHost(hostname).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.InstallPubkey = function (hostname, user, pass, cb) {
        self._proxy.InstallPubkey(hostname, user, pass).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.GetGlobalPolicy = function (cb) {
        self._proxy.GetGlobalPolicy().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.SetGlobalPolicy = function (policy, cb) {
        self._proxy.SetGlobalPolicy(policy).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    // Profile management methods
    this.GetProfiles = function (cb) {
        self._proxy.GetProfiles().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.GetProfile = function (uid, cb) {
        self._proxy.GetProfile(uid).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.DeleteProfile = function (uid, cb) {
        self._proxy.DeleteProfile(uid).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.SaveProfile = function (data, cb) {
        self._proxy.SaveProfile(JSON.stringify(data)).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    // Favourites management methods
    this.HighlightedApps = function (data, uid, cb) {
        self._proxy.HighlightedApps(JSON.stringify(data), uid).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    // Live session methods
    this.ListDomains = function (cb) {
        self._proxy.ListDomains().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.SessionStart = function (uuid, cb) {
        self._proxy.SessionStart(uuid).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.SessionStop = function (cb) {
        self._proxy.SessionStop().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.SessionSave = function (uid, changesets, cb) {
        self._proxy.SessionSave(uid, JSON.stringify(changesets)).done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    this.IsSessionActive = function (uid, cb) {
        uid = uid || '';
        self._proxy.IsSessionActive(uid).done(
            function (resp) {
                cb(resp);
            }
        ).fail(errorhandler);
    };

    // GOA methods
    this.GetGOAProviders = function (cb) {
        self._proxy.GetGOAProviders().done(
            function (resp) {
                cb(JSON.parse(resp));
            }
        ).fail(errorhandler);
    };

    // Quit method
    this.Quit = function (cb) {
        self._proxy.Quit().done(cb).fail(errorhandler);
    };
}

export { FleetCommanderDbusClient };

File: tests/fcdbusclient.py
# -*- coding: utf-8 -*-
# vi:ts=4 sw=4 sts=4

# Copyright (C) 2015 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the licence, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program; if not, see <http://www.gnu.org/licenses/>.
#
# Authors: Alberto Ruiz <aruiz@redhat.com>
#          Oliver Gutiérrez <ogutierrez@redhat.com>

# Python imports
# from __future__ import absolute_import
# import os
# import sys
import time
import json
import logging

import dbus

# PYTHONPATH = os.path.join(os.environ['TOPSRCDIR'], 'admin')
# sys.path.append(PYTHONPATH)

logger = logging.getLogger(__name__)

DBUS_BUS_NAME = "org.freedesktop.FleetCommanderTest"
DBUS_OBJECT_PATH = "/org/freedesktop/FleetCommanderTest"
DBUS_INTERFACE_NAME = "org.freedesktop.FleetCommander"
DBUS_TESTING_INTERFACE_NAME = "org.freedesktop.FleetCommanderTest"


class FleetCommanderDbusClient:

    """
    Fleet commander dbus client
    """

    DEFAULT_BUS = dbus.SessionBus
    CONNECTION_TIMEOUT = 1

    def __init__(self, bus=None):
        """
        Class initialization
        """
        if bus is None:
            bus = self.DEFAULT_BUS()
        self.bus = bus

        t = time.time()
        while time.time() - t < self.CONNECTION_TIMEOUT:
            try:
                self.obj = self.bus.get_object(DBUS_BUS_NAME, DBUS_OBJECT_PATH)
                self.iface = dbus.Interface(
                    self.obj, dbus_interface=DBUS_INTERFACE_NAME
                )
                return
            except Exception:
                pass
        raise Exception("Timed out trying to connect to fleet commander dbus service")

    def get_initial_values(self):
        return self.iface.GetInitialValues()

    def do_domain_connection(self):
        return self.iface.DoDomainConnection()

    def heartbeat(self):
        return self.iface.HeartBeat()

    def check_needs_configuration(self):
        return self.iface.CheckNeedsConfiguration()

    def get_public_key(self):
        return self.iface.GetPublicKey()

    def check_hypervisor_config(self, data):
        return json.loads(self.iface.CheckHypervisorConfig(json.dumps(data)))

    def get_hypervisor_config(self):
        return json.loads(self.iface.GetHypervisorConfig())

    def set_hypervisor_config(self, data):
        return json.loads(self.iface.SetHypervisorConfig(json.dumps(data)))

    def check_known_host(self, host):
        return json.loads(self.iface.CheckKnownHost(host))

    def add_known_host(self, host):
        return json.loads(self.iface.AddKnownHost(host))

    def install_pubkey(self, host, user, passwd):
        return json.loads(self.iface.InstallPubkey(host, user, passwd))

    def get_global_policy(self):
        return json.loads(self.iface.GetGlobalPolicy())

    def set_global_policy(self, policy):
        return json.loads(self.iface.SetGlobalPolicy(policy))

    def save_profile(self, profiledata):
        return json.loads(self.iface.SaveProfile(json.dumps(profiledata)))

    def get_profiles(self):
        return json.loads(self.iface.GetProfiles())

    def get_profile(self, uid):
        return json.loads(self.iface.GetProfile(uid))

    def delete_profile(self, uid):
        return json.loads(self.iface.DeleteProfile(uid))

    def list_domains(self):
        return json.loads(self.iface.ListDomains())

    def session_start(self, domain_uuid):
        return json.loads(self.iface.SessionStart(domain_uuid))

    def session_stop(self):
        return json.loads(self.iface.SessionStop())

    def session_save(self, uid, data):
        return json.loads(self.iface.SessionSave(uid, json.dumps(data)))

    def is_session_active(self, uuid=""):
        return self.iface.IsSessionActive(uuid)

    def get_change_listener_port(self):
        return self.iface.GetChangeListenerPort()

    def get_goa_providers(self):
        return json.loads(self.iface.GetGOAProviders())

    def quit(self):
        return self.iface.Quit()
